src/Microsoft.ML.AutoML/PipelineSuggesters/PipelineSuggester.cs

```diff
@@ -188,30 +191,39 @@ private static IValueGenerator[] ConvertToValueGenerators(IEnumerable<SweepableP
         /// Samples new hyperparameters for the trainer, and sets them.
         /// Returns true if success (new hyperparameters were suggested and set). Else, returns false.
         /// </summary>
-        private static bool SampleHyperparameters(MLContext context, SuggestedTrainer trainer, IEnumerable<SuggestedPipelineRunDetail> history, bool isMaximizingMetric)
+        private static bool SampleHyperparameters(MLContext context, SuggestedTrainer trainer,
+            IEnumerable<SuggestedPipelineRunDetail> history, bool isMaximizingMetric, IChannel logger)
         {
-            var sps = ConvertToValueGenerators(trainer.SweepParams);
-            var sweeper = new SmacSweeper(context,
-                new SmacSweeper.Arguments
+            try
+            {
+                var sps = ConvertToValueGenerators(trainer.SweepParams);
+                var sweeper = new SmacSweeper(context,
+                    new SmacSweeper.Arguments
+                    {
+                        SweptParameters = sps
+                    });
+
+                IEnumerable<SuggestedPipelineRunDetail> historyToUse = history
+                    .Where(r => r.RunSucceeded && r.Pipeline.Trainer.TrainerName == trainer.TrainerName && r.Pipeline.Trainer.HyperParamSet != null && r.Pipeline.Trainer.HyperParamSet.Any());
+
+                // get new set of hyperparameter values
+                var proposedParamSet = sweeper.ProposeSweeps(1, historyToUse.Select(h => h.ToRunResult(isMaximizingMetric))).First();
+                if (!proposedParamSet.Any())
                 {
-                    SweptParameters = sps
-                });
+                    return false;
+                }
 
-            IEnumerable<SuggestedPipelineRunDetail> historyToUse = history
-                .Where(r => r.RunSucceeded && r.Pipeline.Trainer.TrainerName == trainer.TrainerName && r.Pipeline.Trainer.HyperParamSet != null && r.Pipeline.Trainer.HyperParamSet.Any());
+                // associate proposed parameter set with trainer, so that smart hyperparameter
+                // sweepers (like KDO) can map them back.
+                trainer.SetHyperparamValues(proposedParamSet);
 
-            // get new set of hyperparameter values
-            var proposedParamSet = sweeper.ProposeSweeps(1, historyToUse.Select(h => h.ToRunResult(isMaximizingMetric))).First();
```

Comment:
From https://github.com/dotnet/machinelearning/issues/5428#issuecomment-715613915:
> ...another issue with AutoML experiment that after about 70 round of experiment...

Iteration ~70 is interesting as it's when the SMAC sweeper moves from random sweeping (where it explores the hyperparameter space) to a bayesian style, and starts to generate new hyperparameter sets and rank them. Hence why it dies in `sweeper.ProposeSweeps()` at ~70.

Sweeping pattern is:
* One iteration per [trainer for the task](https://github.com/dotnet/machinelearning/blob/e50c4d20012e0d62852f404ae443afca7dad043e/src/Microsoft.ML.AutoML/API/RegressionExperiment.cs#L67-L111) -- Trying defaults for each trainer (~8-12 iterations)
  * Cull to top 3 trainers
* 20 iterations * 3 trainers -- Warm-up random sweeping (60 iterations)
* Until time expires -- SMAC based Bayesian style hyperparameter optimization (hopefully another 50 - 100 iterations)  <= dies at the start of this step

Reply:
Thanks for explaining the behavior of AutoML experiment here.
I debugged the issue and looks like for this case at some stage (around 70 iterations) the Score of all valid history becomes NaN and NaN are not accepted as valid number for Trainer (FastTree in this case) thus we got "All instances skipped due to missing features" error.
Here I filtered the history with additional condition: score must be finite.

---
In reply to: [511429267](https://github.com/dotnet/machinelearning/pull/5455#discussion_r511429267) [](ancestors = 511429267)

Reply:
Interesting. I would have expected that FastTree would simply skip rows with NaN labels.

FastTree may be failing due to NaNs in the feature vector (aka, the hyperparameter set for the pipeline). 

As a test, would it be possible to fill in a value (perhaps `0.5`) for the score field? This would test if (1) SMAC is dying due to the score column being NaN; or (2) our filtering of `FloatUtils.IsFinite(r.Score))`, causes indirect removing rows which would cause SMAC to fail due to missing values within the feature vector.

If filling with `0.5` causes SMAC to work, then your filtering by inspecting the score is a good method. If it still fails, we should instead and filter the history by looking for NaN values in the feature vectors (or perhaps consider throwing as they are unexpected).

NaN values for metrics are expected/semi-common, but I wouldn't expect a NaN in the feature vector. 

After filtering, are there any rows left from the history? Does it continue sweeping, or does it exit at ~70?

Reply:
I would have expected that FastTree would simply skip rows with NaN labels.
 > Yes, this is exactlly what happens here except in some cases all label used are NaN and trainer (FastTree) complaints about that (trainer need at least 1 valid lable). After filtering, sampling method will return at below line and no sampling will happen here: [https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.AutoML/PipelineSuggesters/PipelineSuggester.cs#L207](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.AutoML/PipelineSuggesters/PipelineSuggester.cs#L207)

FastTree may be failing due to NaNs in the feature vector (aka, the hyperparameter set for the pipeline). 
 > As far as I see, feature vector looks good in this case.

As a test, would it be possible to fill in a value (perhaps 0.5) for the score field? This would test if (1) SMAC is dying due to the score column being NaN; or (2) our filtering of FloatUtils.IsFinite(r.Score)), causes indirect removing rows which would cause SMAC to fail due to missing values within the feature vector.
 > Yes, I do some tests as you suggested changing Score from NaN to 0.5 for test purpose and experiment will success running 150 iterations for 10 minutes.

NaN values for metrics are expected/semi-common, but I wouldn't expect a NaN in the feature vector. 
 > NaN values for metrics (label) will be filter out during dataview cursor acessing so filter out in advance should keep logic consistented. [https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Training/TrainerUtils.cs#L856](https://github.com/dotnet/machinelearning/blob/master/src/Microsoft.ML.Data/Training/TrainerUtils.cs#L856)

After filtering, are there any rows left from the history? Does it continue sweeping, or does it exit at ~70?
 >After filtering (instead of change NaN to 0.5), the experiment continue sweeping and return at 150 iteration, history data for these 150 iteration looks valid.



---
In reply to: [513323635](https://github.com/dotnet/machinelearning/pull/5455#discussion_r513323635) [](ancestors = 513323635)

Reply:
@frank-dong-ms: Thanks for investigating. This all sounds good and the NaN filtering on the score does seem like the right route. 

I'd recommend pushing in https://github.com/dotnet/machinelearning/pull/5163 before this PR.

Reply:
@justinormont Sure, thanks Justin for your review and great suggestions.